### PR TITLE
Allow passing of function to .matches() options/message param

### LINKED ISSUE #849: Error message provided via function not being returned in validation errors when using .matches()
**Describe the bug**
As of version 0.28.4 the default error message is returned when a function is provided for the message param in string.matches() rather than the given message within the function

**To Reproduce**
[https://codesandbox.io/s/eager-leftpad-sbpxk?file=/src/index.test.js](https://codesandbox.io/s/eager-leftpad-sbpxk?file=/src/index.test.js)

**Expected behavior**
As per prior versions, expect to pass a function that returns a string and be provided with the given message in the event of an error

**Platform (please complete the following information):**
 - Browser Chrome
 - Version 80

**Additional context**
Aim to provide a PR to fix this

### PR DESCRIPTION
**Pull request by KeelanM90**
Came across issue #849 this morning. Hopefully this closes #849

### CODE REVIEW

**Comment by KeelanM90** [comment]
Please let me know if any updates are needed

File: src/string.js [hunk 1]
```diff
@@ -77,6 +77,7 @@ inherits(StringSchema, MixedSchema, {
 
     if (options) {
       if (typeof options === 'string') message = options;
+      if (typeof options === 'function') message = options;
```
Comment (jquense) [inline src/string.js:80]: I think this whole inner branch can be rewritten as 
```js
if (typeof options === 'object') {
  ({ excludeEmptyString, message, name } = options);
} else {
  message = options
}
```

Comment (KeelanM90) [reply src/string.js:80]: Thanks for the review, I had defaulted to checking it was a function as there was type checking for strings. The docs do show the supported types so it's really down to user error if they pass an incompatible type

Comment (jquense) [reply src/string.js]: in practice, messages can be any type! The string check seems like a bug

Comment (KeelanM90) [reply src/string.js]: Perfect, have pushed the update, thanks again!

**Comment by jquense** [comment]
thanks!

**Pull request merged**